milestones/data.py

```diff
@@ -0,0 +1,373 @@
+
+
+def delete_milestone(milestone):
+    """
+    Deletes an existing milestone from app/local state
+    No return currently defined for this operation
+    """
+    milestone_obj = serializers.deserialize_milestone(milestone)
+    _delete_milestone(milestone_obj)
+
+
+def _delete_milestone(milestone):
+    """
+    Internal helper for milestone removals -- also removes defined dependencies
+    """
+    # Remove related entities, and then remove the Milestone
+    internal.CourseMilestone.objects.filter(
+        milestone=milestone.id).delete()
+    internal.CourseContentMilestone.objects.filter(
+        milestone=milestone.id).delete()
+    internal.UserMilestone.objects.filter(
+        milestone=milestone.id).delete()
+    internal.Milestone.objects.filter(
+        id=milestone.id).delete()
+
+
+def fetch_milestones(milestone):
+    """
+    Retrieves a set of matching milestones from app/local state
+    Returns a list-of-dicts representation of the object
+    """
+    if milestone is None:
+        raise exceptions.InvalidMilestoneException()
+    milestone_obj = serializers.deserialize_milestone(milestone)
+    if milestone_obj.id is not None:
+        return serializers.serialize_milestones(internal.Milestone.objects.filter(
+            id=milestone_obj.id,
+            active=True,
+        ))
+    if milestone_obj.namespace is not None:
+        return serializers.serialize_milestones(internal.Milestone.objects.filter(
+            namespace=milestone_obj.namespace,
+            active=True
+        ))
+    return []
+
+
+def create_course_milestone(course_key, relationship, milestone):
+    """
+    Inserts a new course-milestone into app/local state
+    No response currently defined for this operation
+    """
+    relationship_type = _get_milestone_relationship_type(relationship)
+    milestone_obj = serializers.deserialize_milestone(milestone)
+    internal.CourseMilestone.objects.get_or_create(
+        course_id=unicode(course_key),
+        milestone=milestone_obj,
+        milestone_relationship_type=relationship_type,
+        active=True,
+    )
+
+
+def delete_course_milestone(course_key, milestone):
+    """
+    Removes an existing course-milestone from app/local state
+    No response currently defined for this operation
+    """
+    try:
+        internal.CourseMilestone.objects.get(
+            course_id=unicode(course_key),
+            milestone=milestone['id'],
+            active=True,
+        ).delete()
+    except internal.CourseMilestone.DoesNotExist:
+        pass
+
+
+def fetch_courses_milestones(course_keys, relationship=None, user=None):
+    """
+    Retrieves the set of milestones currently linked to the specified courses
+    Optionally pass in 'relationship' (ex. 'fulfills') to filter down the set
+    """
+    queryset = internal.CourseMilestone.objects.filter(
+        course_id__in=course_keys,
+        active=True
+    ).select_related('milestone')
```

Comment:
On that note, could you please add `assertNumQueries` tests? We've got a lot of joins going on -- it'd be easy to accidentally introduce performance regressions without a test safety net.